src/Elastic.Apm/Model/Transaction.cs

```diff
@@ -70,12 +70,25 @@ ICurrentExecutionSegmentsContainer currentExecutionSegmentsContainer
 			HasCustomName = false;
 			Type = type;
 
+			StartActivity();
+
 			var isSamplingFromDistributedTracingData = false;
 			if (distributedTracingData == null)
 			{
-				var traceIdBytes = new byte[16];
-				TraceId = RandomGenerator.GenerateRandomBytesAsString(traceIdBytes);
+				// Here we ignore Activity.Current.ActivityTraceFlags because it starts out without setting the IsSampled flag, so relying on that would mean a transaction is never sampled.
 				IsSampled = sampler.DecideIfToSample(idBytes);
+
+				if (Activity.Current != null && Activity.Current.IdFormat == ActivityIdFormat.W3C)
```

Comment:
It might be worth considering that this change breaks backward compatibility - before this change transactions started using public API with `null` `DistributedTracingData` created new trace ID.

Reply:
I'm not sure this is a problem and I'm also not sure I'd consider this as breaking backward compatibility.

I think this only means that prior to this a new TraceId was generated and now the agent picks up the TraceId from Activity. I think this only means the agent ignored the TraceId on the Activity previously but now it does not - I'm not sure this is something a user would rely on.

Or is there anything else I miss here?

Reply:
I cannot say for sure that there are real users that might rely on public API generating different trace IDs for transactions started using public API with `null` `DistributedTracingData`.

I can imagine a scenario where an application reads a batch of tasks from somewhere and kicks off distributed processing of each task. Before it was possible to start a separate trace for each task but now if this processing happens in the context of an existing Activity (which might have been created by some library over which user doesn't have any control and doesn't care about Activities it creates) it's not possible to start a separate trace for each task anymore.

One way to solve this issue is to make this change as opt-in - agent can opt-in for transactions it creates automatically but public API will not opt-in by default to keep backward compatibility.

Reply:
Oh, I see the use-case you mean. Definitely a valid concern and indeed in that case the 2 transactions would be on the same trace now.

But why is this a backward compatibility issue? 🤔 Definitely changes how the agent behaves - and after the update the transactions will be part of the same trace (unlike before), but to me the question is more like if this is considered a breaking change or not.

Regarding making it opt-in: could be an option, but not sure we need to go into that complication.

One thing to consider here is that this only changes the behaviour if there is an active transaction and the user starts multiple separate transactions within the same chain of activities. E.g. in case of an ASP.NET Core app that would mean 2 transactions within the same HTTP request - and similar examples. One could argue that in this case making those multiple transactions as part of the same trace is actually correct and not having them in the same trace would be incorrect.

So usually when a library starts an activity it also ends it when the "unit of work" (like an HTTP request) is finished (and with that basically "resets" the TraceId). So creating multiple transactions with the same Activity really seems like they should be in the same trace.

To me this feels like something we could accept - but we can follow up on this in a planning.